pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/BitmapInvertedIndexWriter.java

```diff
@@ -0,0 +1,126 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.segment.creator.impl.inv;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.FileChannel;
+import org.apache.pinot.segment.spi.memory.CleanerUtil;
+import org.roaringbitmap.RoaringBitmap;
+
+
+/**
+ * Writer for bitmap inverted index file.
+ * <pre>
+ * Layout for RoaringBitmap inverted index:
+ * |-------------------------------------------------------------------------|
+ * |                    Start offset of 1st bitmap                           |
+ * |    End offset of 1st bitmap (exclusive) / Start offset of 2nd bitmap    |
+ * |                                   ...                                   |
+ * | End offset of 2nd last bitmap (exclusive) / Start offset of last bitmap |
+ * |                  End offset of last bitmap (exclusive)                  |
+ * |-------------------------------------------------------------------------|
+ * |                           Data for 1st bitmap                           |
+ * |                           Data for 2nd bitmap                           |
+ * |                                   ...                                   |
+ * |                           Data for last bitmap                          |
+ * |-------------------------------------------------------------------------|
+ * </pre>
+ */
+public final class BitmapInvertedIndexWriter implements Closeable {
+  // 264MB - worst case serialized size of a single bitmap with Integer.MAX_VALUE rows
+  private static final long MAX_INITIAL_BUFFER_SIZE = 256 << 20;
+  // 128KB derived from 1M rows (15 containers), worst case 8KB per container = 120KB + 8KB extra
+  private static final long PESSIMISTIC_BITMAP_SIZE_ESTIMATE = 128 << 10;
+  private final FileChannel _fileChannel;
+  private final ByteBuffer _offsetBuffer;
+  private ByteBuffer _bitmapBuffer;
+  private int _bytesWritten;
+
+  public BitmapInvertedIndexWriter(File outputFile, int numBitmaps)
+      throws IOException {
+    int sizeForOffsets = (numBitmaps + 1) * Integer.BYTES;
+    long bitmapBufferEstimate = Math.min(PESSIMISTIC_BITMAP_SIZE_ESTIMATE * numBitmaps, MAX_INITIAL_BUFFER_SIZE);
+    _fileChannel = new RandomAccessFile(outputFile, "rw").getChannel();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource that might leak to a class field, causing a resource leak that slows down or crashes your system. For example, when FileChannel.map throws an exception, the *BitmapInvertedIndexWriter* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *\_fileChannel* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).